LIPS/lip-composite_beacon_report_receiver.md

```diff
@@ -0,0 +1,158 @@
+event CallbackAdded(address indexed callback, uint256 atIndex);
+```
+Emitted when a new `callback` is added at the `atIndex` position.
+See: `insertCallback`, `removeCallback`.
+
+```solidity
+event CallbackRemoved(address indexed callback, uint256 atIndex);
+```
+Emitted when a `callback` was removed from the `atIndex` position.
+See: `removeCallbacks`.
+
+### `CompositePostRebaseBeaconReceiver` contract
+
+The contract inherited from `OrderedCallbacksArray` to implement a composite design pattern applicable for the `LidoOracle` contract as `IBeaconReportReceiver`.
+
+```solidity
+modifier onlyOracle()
+```
+Enforces the `msg.sender` to equal `oracle` address which is set upon the contract construction.
+
+```solidity
+constructor(address _voting, address _oracle) OrderedCallbacksArray(_voting);
+```
+Init contract and store the provided `_voting` and `_oracle` addresses to arm the `onlyVoting` and `onlyOracle` modifiers.
+See: `OrderedCallbacksArray.onlyVoting`, `onlyOracle`.
+
+```solidity
+function processLidoOracleReport(
+    uint256 _postTotalPooledEther,
+    uint256 _preTotalPooledEther,
+    uint256 _timeElapsed
+) external override onlyOracle
+```
+Implements the `IBeaconReceiver` interface, which is suitable for the `LidoOracle` contract.
+Iteratively calls `callback.processLidoOracleReport` for each stored `callback` in the callbacks array preserving the order.
+
+## Backward compatibility
+
+* The proposed `CompositePostRebaseBeaconReceiver` could be used with only one `IBeaconReportReceiver` callback inside, providing the same behavior as before by using the wrapped callback directly (without wrapping into a composite receiver).
+* Even if the callbacks array is empty, the existing behavior preserves, i.e., there are no additional side-effects incurred for the lido oracle reports except the gas spendings.
+
+## Gas effects
+
+In the case of `ComositePostRebaseBeaconReceiver` used with only one `IBeaconReportReceiver` callback set, spendings increase approximately by the ~4600 gas compared to direct usage of the underlying callback with `LidoOracle` (without composite adapter).
+
+## Security considerations
+
+#### Upgradability
+The proposed contracts are non-upgradable for the sake of simplicity. In case of emergency, we can disable them entirely by calling `LidoOracle.setBeaconReportReceiver(address(0))` and redeploy newly developed versions later from scratch without incurring additional proxy-initialization logic and state transitions.
+
+#### Access control (permissions model)
+There are two permissioned addresses introduced explicitly: `Voting` and `LidoOracle`. Only `Voting` can add/insert/remove callbacks into the `CompositePostReabaseBeaconReceiver` to make it feasible only by the Lido DAO direct vote-backed will. Only `LidoOracle` is allowed to call `processLidoOracleReport` to prevent propagating various fake oracle reports by the added callbacks or execution in an arbitrary time moment.
+
+#### Misbehaving callback
```

Comment:
I propose to note here that if a callback throws an exception all the subsequent callbacks get skipped.

Reply:
Good point, I'll add it.